### PR TITLE
Allow adding custom loopback client in acquireTokenInteractive

### LINKED ISSUE #5442: Problem with ElectronSystemBrowserTestApp deep linking with acquireTokenInteractive()
### Core Library

MSAL Node (@azure/msal-node)

### Core Library Version

1.14.2

### Wrapper Library

Not Applicable

### Wrapper Library Version

none

### Public or Confidential Client?

Public

### Description

I have the [ElectronSystemBrowserTestApp ](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-node-samples/ElectronSystemBrowserTestApp) sample app setup and running, and can sign and the deeplink calls back to the app. The problem is how to improve the experience with the browser popup to allow the redirect. Since acquireTokenInteractive chooses a random port on localhost each time, the user will always have the browser popup, even if they check the box to allow to open link, because the next time it will be a different port on localhost.
image

### Error Message

_No response_

### Msal Logs

_No response_

### MSAL Configuration

```javascript
{
    "authOptions":
        {
            "clientId": "Enter_the_Application_Id_Here",
            "authority": "https://login.microsoftonline.com/organizations"
        },
    "resourceApi":
    {
        "endpoint": "Enter_the_Graph_Endpoint_Here",
        "scopes": ["User.Read", "Mail.Read"]
    },
    "customProtocol":
    {
        "name": "msal{Enter_the_Application_Id_Here}"
    },
    "cache": {
        "cacheLocation": "./data/cache.json"
    }
}
```

### Relevant Code Snippets

```javascript
ElectronSystemBrowserTestApp: https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-node-samples/ElectronSystemBrowserTestApp
```

### Reproduction Steps

1. Download the [ElectronSystemBrowserTestApp ](https://github.dev/AzureAD/microsoft-authentication-library-for-js/blob/dev/samples/msal-node-samples/ElectronSystemBrowserTestApp/src/AuthProvider.ts) and configure the application in customConfig.js and enable deeplinking
2. Run the app and sign in > browser will open and after sign in is complete will try to deeplink back to the app, but browser will popup to allow to open this applicaiton.
3. The problem is the browser will popup every time, even if the users checks always allow, because acquireTokenInteractive uses a different port every time.

### Expected Behavior

Expected behavior would be some way to specify a port number or a way to have acquireTokenInteractive use a consistent URL so that the user would not always get the popup consent in browser for the deep link back to the electron app****

### Identity Provider

Azure AD / MSA

### Browsers Affected (Select all that apply)

Chrome, Firefox, Edge, Safari, Internet Explorer

### Regression

_No response_

### Source

External (Customer)

### PR DESCRIPTION
**Pull request by derisen**
This PR adds a loopback client interface to `InteractiveRequest` for devs who would like to run their own server for listening the authZ code response in `acquireTokenInteractive` API.

### CODE REVIEW

**Review by peterzenz — APPROVED** [review]

File: lib/msal-node/src/client/PublicClientApplication.ts [hunk 1]
```diff
@@ -90,9 +91,16 @@ export class PublicClientApplication extends ClientApplication implements IPubli
...
-        const { openBrowser, successTemplate, errorTemplate, ...remainingProperties } = request;
+        const { openBrowser, successTemplate, errorTemplate, customLoopbackClient, ...remainingProperties } = request;
+
+        let loopbackClient: ILoopbackClient;
+
+        if (customLoopbackClient) {
+            loopbackClient = customLoopbackClient;
+        } else {
+            loopbackClient = new LoopbackClient();
+        }
```
Comment (tnorling) [inline lib/msal-node/src/client/PublicClientApplication.ts]:
```suggestion
        const loopbackClient: ILoopbackClient = customLoopbackClient || new LoopbackClient();
```

File: lib/msal-node/src/request/InteractiveRequest.ts [hunk 2]
```diff
@@ -3,11 +3,13 @@
...
 
+import { ILoopbackClient } from "../network/ILoopbackClient";
 import { AuthorizationUrlRequest } from "./AuthorizationUrlRequest";
 
 export type InteractiveRequest = Pick<AuthorizationUrlRequest, "authority"|"correlationId"|"claims"|"azureCloudOptions"|"account"|"extraQueryParameters"|"tokenQueryParameters"|"extraScopesToConsent"|"loginHint"|"prompt"> & {
     openBrowser: (url: string) => Promise<void>;
     scopes?: Array<string>;
     successTemplate?: string;
     errorTemplate?: string;
+    customLoopbackClient?: ILoopbackClient
```
Comment (tnorling) [inline lib/msal-node/src/request/InteractiveRequest.ts]: Let's just call this loopbackClient
```suggestion
    loopbackClient?: ILoopbackClient
```

**Review by tnorling — COMMENTED** [review]
Looks good, 2 nits and any docs or samples we can add to help illustrate usage?

File: lib/msal-node/src/request/InteractiveRequest.ts (hunk 2 above)
Comment (derisen) [reply lib/msal-node/src/request/InteractiveRequest.ts]: done!

File: lib/msal-node/src/client/PublicClientApplication.ts [hunk 3]
```diff
@@ -86,13 +87,14 @@ export class PublicClientApplication extends ClientApplication implements IPubli
     }
 
     /**
-     * Acquires a token by requesting an Authorization code then exchanging it for a token.
+     * Acquires a token interactively via the browser by requesting an authorization code then exchanging it for a token.
      */
-    async acquireTokenInteractive(request: InteractiveRequest): Promise<AuthenticationResult> {
+    public async acquireTokenInteractive(request: InteractiveRequest): Promise<AuthenticationResult> {
```
Comment (tnorling) [inline lib/msal-node/src/client/PublicClientApplication.ts:92]: nit: not needed
```suggestion
    async acquireTokenInteractive(request: InteractiveRequest): Promise<AuthenticationResult> {
```

**Review by tnorling — APPROVED** [review]
Looks good, thanks!

File: lib/msal-node/src/client/PublicClientApplication.ts (hunk 3 above)
Comment (derisen) [reply lib/msal-node/src/client/PublicClientApplication.ts:92]: true, but given the rest of `acquireToken*` implementations has the modifier explicitly, shouldn't we keep this for consistency?

**Comment by derisen** [comment]
> Looks good, 2 nits and any docs or samples we can add to help illustrate usage?

done and done!

**Review by sameerag — APPROVED** [review]
Looks good. Do we have a docs (and .md file) defining why someone would need a loop back server? If yes, please link here, if not, can we add it? Can be handled as a separate PR.

**Comment by derisen** [comment]
> Looks good. Do we have a docs (and .md file) defining why someone would need a loop back server? If yes, please link here, if not, can we add it? Can be handled as a separate PR.

Thanks! Added comments for API documentation and also to an Electron sample for usage. Need to update/re-organize some of our .md docs for API usage -so probably best to tackle this separately in another PR

**Pull request merged**